crates/ra_editor/src/typing.rs

```diff
@@ -140,30 +141,71 @@ fn remove_newline(
     offset: TextUnit,
 ) {
     if node.kind() == WHITESPACE && node_text.bytes().filter(|&b| b == b'\n').count() == 1 {
+        // Special case that turns something like:
+        //
+        // ```
+        // my_function({<|>
+        //    <some-expr>
+        // })
+        // ```
+        //
+        // into `my_function(<some-expr>)`
         if join_single_expr_block(edit, node).is_some() {
             return
         }
-        match (node.prev_sibling(), node.next_sibling()) {
-            (Some(prev), Some(next)) => {
-                let range = TextRange::from_to(prev.range().start(), node.range().end());
-                if is_trailing_comma(prev.kind(), next.kind()) {
-                    edit.delete(range);
-                } else if no_space_required(prev.kind(), next.kind()) {
-                    edit.delete(node.range());
-                } else if prev.kind() == COMMA && next.kind() == R_CURLY {
-                    edit.replace(range, " ".to_string());
+
+        if let (Some(prev), Some(next)) = (node.prev_sibling(), node.next_sibling()) {
+            let range = TextRange::from_to(prev.range().start(), node.range().end());
+            if is_trailing_comma(prev.kind(), next.kind()) {
+                // Removes: trailing comma, newline (incl. surrounding whitespace)
+                edit.delete(range);
+            } else if no_space_required(prev.kind(), next.kind()) {
```

Comment:
I wonder if we should use [uncover](https://crates.io/crates/uncover) ([blog post](https://matklad.github.io/2018/06/18/a-trick-for-test-maintenance.html)) here, to easily connect all this special-cases with relevant tests.